src/data/list/infix.lean

```diff
@@ -208,24 +218,36 @@ lemma suffix_iff_eq_drop : l₁ <:+ l₂ ↔ l₁ = drop (length l₂ - length l
  λ e, e.symm ▸ drop_suffix _ _⟩
 
 instance decidable_prefix [decidable_eq α] : ∀ (l₁ l₂ : list α), decidable (l₁ <+: l₂)
-| []      l₂ := is_true ⟨l₂, rfl⟩
-| (a :: l₁) [] := is_false $ λ ⟨t, te⟩, list.no_confusion te
+| []        l₂        := is_true ⟨l₂, rfl⟩
+| (a :: l₁) []        := is_false $ λ ⟨t, te⟩, list.no_confusion te
 | (a :: l₁) (b :: l₂) :=
   if h : a = b then
-    @decidable_of_iff _ _ (by rw [← h, prefix_cons_inj])
-      (decidable_prefix l₁ l₂)
+    decidable_of_decidable_of_iff (decidable_prefix l₁ l₂) (by rw [← h, prefix_cons_inj])
   else
     is_false $ λ ⟨t, te⟩, h $ by injection te
 
 -- Alternatively, use mem_tails
 instance decidable_suffix [decidable_eq α] : ∀ (l₁ l₂ : list α), decidable (l₁ <:+ l₂)
-| []      l₂ := is_true ⟨l₂, append_nil _⟩
+| []        l₂ := is_true ⟨l₂, append_nil _⟩
 | (a :: l₁) [] := is_false $ mt (length_le_of_sublist ∘ is_suffix.sublist) dec_trivial
-| l₁      l₂ := let len1 := length l₁, len2 := length l₂ in
+| l₁        l₂ := let len1 := length l₁, len2 := length l₂ in
   if hl : len1 ≤ len2 then
     decidable_of_iff' (l₁ = drop (len2-len1) l₂) suffix_iff_eq_drop
   else is_false $ λ h, hl $ length_le_of_sublist $ h.sublist
 
+instance decidable_infix [decidable_eq α] : ∀ (l₁ l₂ : list α), decidable (l₁ <:+: l₂)
+| []        l₂        := is_true ⟨[], l₂, rfl⟩
+| (a :: l₁) []        := is_false $ λ ⟨s, t, te⟩, by simp at te; exact te
+| l₁        (b :: l₂) :=
+  if h : l₁ <+: b :: l₂ then
+    is_true (match h with ⟨t, ht⟩ := ⟨[], t, nil_append l₁ ▸ ht⟩ end)
+  else
+    decidable_of_decidable_of_iff (decidable_infix l₁ l₂)
+      ⟨infix_cons, λ h', match h' with
+      | ⟨[]     , t, e⟩ := by simp at e; exact (h ⟨t, e⟩).elim
+      | ⟨c :: cs, t, e⟩ := ⟨cs, t, tail_eq_of_cons_eq e⟩
+      end⟩
```

Comment:
What's the motivation for moving this lemma and making the proof longer?

Reply:
`decidable_infix` was in `section inits_tails`, while `decidable_prefix` and `decidable_suffix` are in `section fix`. It seems that they should be put together in `section fix`.

Reply:
Now it's shorter.